### ISSUE TITLE
[Workstream A] PR 5 of 8: Embedding and reranker bake-off on CPU, resulting in ADR-0001

### ISSUE THREAD

**Issue by zaxbysauce** [ZaxbyHub/trainingapp#55]
**Roadmap:** A5 · **Workstream A: Contract, evidence and decisions**
**Sequence:** PR 5 of 8 (tracking slot; dependencies define execution order)
**Roadmap baseline:** 2026-09-05 · **Priority:** P1 · **Repo revision:** b2ddda06e86b63f81bdbedce220a3a2357bc9852

## Outcome
When this closes, there is a merged ADR naming one embedding model and one reranker model for ALL surfaces (desktop Python, browser web_ui, future Electron/Node), backed by measured quality (on the A4 eval set) and measured CPU cost. This issue does NOT implement the switch in either surface's production code — downstream issues (Electron's WS-B PR 7/#65, and any desktop-side follow-up) consume the ADR's decision. Because there is no deployed embedding index depending on backward compatibility yet (per implementation_plan.md §4: "There are no deployed clients, so there is no compatibility constraint"), this bake-off is free to pick a different model than either surface currently uses.

## Evidence motivating this slot
- CONFIRMED desktop uses BAAI/bge-small-en-v1.5, ~129 MB on disk, 384-dim (audit_report.md §1.2, cross-referenced against `models/bge-small-en-v1.5` directory name).
- CONFIRMED web_ui uses snowflake-arctic-embed-m-v1.5 q8, 768-dim via Transformers.js (audit_report.md §2.1, citing `embedding-service.ts:9,28-30,160-201`).
- CONFIRMED the desktop reranker is inert on shipped installs: `reranking.py:62` uses `local_files_only=True`; only `scripts/precache_models.py:35` fetches the model and nothing but its own test invokes that script; `DocumentQAApp.spec:5` and `AFOMIS.spec:19` ship `models/` but not the HuggingFace cache — so the configured cross-encoder (`cross-encoder/ms-marco-MiniLM-L6-v2`, itself not even the canonical HF id per audit_report.md line 37, which is `...MiniLM-L-6-v2`) silently fails to load and retrieval falls back to plain top-k (caught at `rag_engine.py:499-510`).
- CONFIRMED web_ui's reranker is `ettin-reranker-32m-v1` (`manifest.json:33` per audit_report.md line 60) and its relevance floor `MIN_CROSS_SCORE=0.2` is explicitly documented as un-revalidated at `rag-orchestrator.ts:148-163`.
- RESEARCH (to-be-measured): EmbeddingGemma-300M (768-dim, Matryoshka-truncatable, ONNX) and Qwen3-Embedding-0.6B are candidate 2026-era embedding models per the audit's recommendations section; arxiv Qwen3 Embedding and sbert.net cross-encoder efficiency are the cited sources for reranker sizing guidance — none of these numbers are measured in this repo yet.

## Prerequisites and ownership
Depends on #54 (A4) — scored on the A4 eval set; running before it exists has no quality metric to optimize against. Parallel-safe with #56. Owns: bake-off methodology, candidate list, CPU-cost measurement, `docs/adr/0001-embedding-reranker.md`. Does not own: implementing the winning model in production code (downstream issues cite "per ADR-0001"), nor fixing the desktop reranker's `local_files_only` packaging defect (name it as a follow-up, don't fix it here).

## Required scope
- [ ] Candidate embeddings: `arctic-embed-m v1.5` (control, web_ui), `EmbeddingGemma-300M`, `Qwen3-Embedding-0.6B`, `bge-small-en-v1.5` (control, desktop).
- [ ] Candidate rerankers: `ettin-reranker-32m-v1` (control), `ms-marco-MiniLM-L-6-v2` (fix the canonical HF id's missing hyphen regardless of winner), `bge-reranker-v2-m3`.
- [ ] Quality scoring: each combination through the A4 harness (#54), reporting recall@k/MRR.
- [ ] CPU cost: reuse/extend #52's ONNX microbenchmark at top-15/top-30 candidate counts.
- [ ] License check per candidate, feeding A6/#56's license gate.
- [ ] `docs/adr/0001-embedding-reranker.md`: decision table (model, dims, quality, CPU cost, license), explicit winner, and a named follow-up if the desktop reranker's `local_files_only`/HF-cache defect (`reranking.py:62`) is unresolved.
- [ ] Record the winning model's `{model_id, dims}` in the ADR for C1/#68 to reference (not this issue's job to build the pack schema).

Out of scope: implementing chosen models in production code; fixing the reranker packaging defect's code; the SQLite pack format (#68).

## Acceptance
| Case | Required proof |
|---|---|
| All 4 embedding candidates scored on the A4 eval set | ADR decision table has a recall@k/MRR row for each of arctic-embed-m, EmbeddingGemma-300M, Qwen3-Embedding-0.6B, bge-small-en-v1.5 |
| All 3 reranker candidates scored | ADR decision table has a recall@k/MRR row for each of ettin-32m, ms-marco-MiniLM-L-6-v2 (corrected id), bge-reranker-v2-m3 |
| CPU cost measured for every candidate | ADR table includes per-candidate latency at top-15/top-30 from the ONNX microbenchmark |
| License recorded for every candidate | ADR table includes license type per model |
| ADR merged with a named winner | `docs/adr/0001-embedding-reranker.md` exists with an explicit "Decision" section naming one embedding model and one reranker for all surfaces |
| Desktop reranker packaging defect is named as a follow-up if unresolved | ADR's follow-up section explicitly references `reranking.py:62` and the missing HF-cache bundling, or links to a tracking issue |

## Evidence and exit gate
Closing PR must attach `docs/adr/0001-embedding-reranker.md` and the underlying eval reports (linked or embedded) for every candidate. Evidence is invalidated if A4's eval set changes materially after this bake-off runs — a significant question-set revision requires re-running the bake-off's quality scoring, not just carrying forward old numbers.

## Execution and closure contract
Treat PR n of total as a stable tracking slot, not merge order. The explicit dependencies govern execution; focused implementation issues supply evidence to later qualification issues without depending on those later issues closing. Re-anchor source and active-PR overlap before implementation. Close only the fully satisfied scope: wired production callers, meaningful focused validation and documented outcomes, with no silent deferral. Open one PR per issue titled with this issue's prefix; keep `contracts/` (API OpenAPI, store schema, pack schema) authoritative; add tests for every acceptance row; update docs touched by the change. Verification-only residuals may close with concrete current evidence rather than an unnecessary code patch.

## Complete workstream execution index

Slots identify work; they do not override dependencies. Totals are authoritative for the v3 roadmap (baseline 2026-09-05). Epic: #50.

### Workstream A: Contract, evidence and decisions (8 slots)

| Slot | Issue | Work |
|---|---|---|
| A1 | #51 | Repair /ask/stream, freeze the API contract, and lock down default bind/auth |
| A2 | #52 | Build a benchmark harness and record a measured perf baseline on the reference i5 laptop |
| A3 | #53 | Fix the desktop RAM gate, raise thread defaults, and surface the real LLM-load diagnostic |
| A4 | #54 | Build the tier-0 eval set and a backend-agnostic retrieval/answer eval harness |
| A5 | #55 | Embedding and reranker bake-off on CPU, resulting in ADR-0001 |
| A6 | #56 | Select distributable LLM profiles by quality, latency, memory, and license, resulting in ADR-0002 |
| A7 | #57 | Spike — prove the packaged native stack and choose Node backend vs Electron-hosted Python sidecar (ADR-0003) |
| A8 | #58 | Spike — prove Storyline embedding, current-slide reads, and exact slide jumps |

### Workstream B: Electron shell and desktop backend (9 slots)

| Slot | Issue | Work |
|---|---|---|
| B1 | #59 | Electron scaffold with secure defaults and CI installer artifact |
| B2 | #60 | Secure the Electron protocol and loopback API transport |
| B3 | #61 | Backend host in Electron implementing the frozen API contract, per ADR-0003 |
| B4 | #62 | Native LLM inference with Quality/Fast profiles, streaming, cancel, resident model |
| B5 | #63 | Prove Node/Python sqlite-vec interoperability and freeze the store schema |
| B6 | #64 | Profile-scoped ingestion, migrations, backup and recovery on the SQLite store |
| B7 | #65 | Hybrid retrieval (sqlite-vec + FTS5 -> RRF -> reranker) with eval parity |
| B8 | #66 | Enforce a 16 GB runtime memory and concurrency budget |
| B9 | #67 | Renderer integration — move Documents/ingest/settings to the API, Electron-aware mode, model/status UI |

### Workstream C: Knowledge packs and freshness (9 slots)

| Slot | Issue | Work |
|---|---|---|
| C1 | #68 | Define the Knowledge Pack specification, JSON Schema, fixtures, validator, and ADR |
| C2 | #69 | Python PackManager lifecycle (install/supersede/rollback/remove) with content-hash chunk ids and update-in-place |
| C3 | #70 | Node PackManager lifecycle on the SQLite store, sharing C2's semantics and test vectors |
| C4 | #71 | Version precedence, cross-pack chunk dedup, and recency-weighted ranking |
| C5 | #72 | Grounded/general provenance field in both APIs and a visible chat UI badge |
| C6 | #73 | packtool CLI — build-docs, verify, diff, with prebuilt embeddings and FTS on a build machine |
| C7 | #74 | Documents page — packs list, versions, drag-drop install, remove/rollback, per-source citations |
| C8 | #75 | Harden pack installation against path traversal, zip bombs, tampering, and version mismatch |
| C9 | #76 | Knowledge packs on the secondary browser surface — ADR deciding a browser adapter vs an explicit capability gate |

### Workstream D: Training pack and Learn panel (7 slots)

| Slot | Issue | Work |
|---|---|---|
| D1 | #77 | Storyline extractor — data.js/frame.js/per-slide JSON to one document per slide + course outline |
| D2 | #78 | Offline narration transcription with content-hash cache (faster-whisper, build machine only) |
| D3 | #79 | Build an installable Storyline pack with a prebuilt training index (packtool build-storyline) |
| D4 | #80 | Recompute doc-to-slide links whenever active packs change |
| D5 | #81 | Embedded Storyline player service + slide-state events (per the A8 recipe) |
| D6 | #82 | Learn panel — answer + "where to learn this" with Open-in-training deep links (both backends) |
| D7 | #83 | Ask-about-this-slide — pinned slide context from the player into chat |

### Workstream E: Hardening, CI and release (6 slots)

| Slot | Issue | Work |
|---|---|---|
| E1 | #84 | Package models and default packs into the Electron installer with an integrity manifest and a <=7 GB size budget |
| E2 | #85 | First-run validation wizard (RAM to profile selection, model presence, pack activation) and licensing notices |
| E3 | #86 | Offline and low-RAM validation matrix on the reference laptop |
| E4 | #87 | Path-filtered CI workflows, conformance on both backends, perf thresholds, pack fixtures, and eval reporting |
| E5 | #88 | Signed application and knowledge-pack update channels with rollback |
| E6 | #89 | Rewrite v3 architecture docs, pack authoring guide, training-pack refresh runbook, CHANGELOG, and retire stale documentation |

**Comment by zaxbysauce** [comment]
**issue-tracer red checkpoint anchor (Phase 2.5, pre-implementation)**

Frozen acceptance checks C1-C7 for the A5 bake-off (4 embedding + 3 reranker candidates, CPU 8-thread cap, ADR-0001). Reviewers: retain this literal and verify with `repro-check.sh verify-anchor` after implementation.

```
issue-tracer-checkpoint-v1 slug=55-embedding-reranker-bakeoff manifest=30ce70c1ccb5507e20d625259010f02e99070fe7 semantics=2e85f671c451647d49a3770c65f3e96af8d1c8ea tree=8a9abbdbba529d77791753a9295e10db068a4cd4
```

**Comment by zaxbysauce** [comment]
**A5 bake-off implemented in PR #124** — https://github.com/ZaxbyHub/trainingapp/pull/124

What changed: `eval/bakeoff/` measurement package (quality grid + 8-thread CPU cost + license provenance for all 4 embedding and 3 reranker candidates on the A4 tier-0 set) and `docs/adr/0001-embedding-reranker.md`, which names **`Qwen/Qwen3-Embedding-0.6B` (dims: 1024)** and **`cross-encoder/ettin-reranker-32m-v1`** for all surfaces, with the measured decision table, consequences, and follow-ups.

How to use it: downstream re-pin issues cite "per ADR-0001"; the winning `{model_id, dims}` for the pack schema (#68) is in the ADR Decision section; reproduction commands are in `eval/bakeoff/README.md`.

Notable findings recorded in the ADR: (1) the shipped `models/ettin-reranker-32m-v1/onnx/*` artifacts used by the browser and Electron rerankers emit logits from an untrained head — both relevance floors are calibrated against noise; (2) the configured `cross-encoder/ms-marco-MiniLM-L6-v2` id IS the canonical HF id (the audit report's hyphen claim was backwards); (3) no ONNX export of ettin anywhere carries its trained head — a correct export is a downstream obligation.

No migration required by this PR itself (no production model changes — that is the downstream issues' scope).

Trace: `.agents/issue-traces/55-embedding-reranker-bakeoff/` (frozen checks C1-C7, anchor receipt published pre-implementation in this issue, plan critic / implementation reviewer / final critic all APPROVE).

**Comment by zaxbysauce** [comment]
**PR #124 merged** (squash de0c5d3987a7f6e9f46b1a0dd21ee8f35cc4822c, all 8 CI checks green on head bf75d6f). Issue #55 / ADR-0001 is closed end to end.

Post-publication review round summary: full swarm-pr-review (6 base lanes + 6 micro risk families) validated 21 findings — including both Copilot review threads (the latent left-padding pooling branch and the TAB-broken fallback path) plus a low-severity ADR range correction — fixed across commits 30cf238/c3f0bd2/bf75d6f, re-gated (fix reviewer APPROVE; final critic NEEDS_REVISION -> 3 closure gaps fixed -> delta re-critic APPROVE), and merged. One CI flake (the documented Windows 3.10 native-crash signature) went green on rerun. Closure ledger: .zcode/pr-review/pr124-20260920/closure-ledger.md.

No production model changes shipped here — the ADR-0001 re-pins (Qwen3-Embedding-0.6B dims 1024 + trained-head ettin export, floor recalibration, 1024-dim store widening) remain with the downstream "per ADR-0001" issues.
